design-proposals/instancetype-vm-rollout-support.md

```diff
@@ -0,0 +1,158 @@
+
+```
+$ # $kind is optional and defaults to virtualmachineclusterinstancetype
+$ virtctl vm update-instance-type $vm $kind/$new-instance-type
+
+$ # $kind is optional and  defaults to virtualmachineclusterpreference
+$ virtctl vm update-preference $vm $kind/$new-preference
+```
+
+This subcommands will provide an optional `--wait/-w` switch that will wait until either a `RestartRequired` condition appears on the VM *or* for the `liveUpdate` process to complete and for all resources to be plugged into the running VMI.
+
+## API Examples
+
+```yaml
+./cluster-up/kubectl.sh apply -k https://github.com/kubevirt/common-instancetypes
+
+./cluster-up/kubectl.sh patch kv/kubevirt -n kubevirt --type merge -p '{"spec":{"workloadUpdateStrategy":{"workloadUpdateMethods":["LiveMigrate"]},"configuration":{"developerConfiguration":{"featureGates": ["VMLiveUpdateFeatures"]},"vmRolloutStrategy": "LiveUpdate", "liveUpdateConfiguration": {"maxGuest": "8Gi"}}}}'
+
+./cluster-up/kubectl.sh apply -f - <<EOF
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: alpine
+spec:
+  instancetype:
+   name: u1.nano
+  runStrategy: Always
+  template:
+    metadata:
+      creationTimestamp: null
+    spec:
+      domain:
+        devices:
+          interfaces:
+          - masquerade: {}
+            name: default
+        machine:
+        resources: {}
+      networks:
+      - name: default
+        pod: {}
+      volumes:
+      - containerDisk:
+          image: registry:5000/kubevirt/alpine-container-disk-demo:devel
+        name: alpine
+      - cloudInitNoCloud:
+          userData: |
+            #!/bin/sh
+            echo 'printed from cloud-init userdata'
+        name: cloudinitdisk
+EOF
+
+./cluster-up/kubectl.sh wait vms/alpine --for=condition=Ready
+
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.spec.domain.memory.guest}'="512Mi"
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.status.memory.guestAtBoot}'="512Mi"
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.status.memory.guestRequested}'="512Mi"
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.status.memory.guestCurrent}'="512Mi"
+
+./cluster-up/kubectl.sh patch vms/alpine --type merge -p '{"spec":{"instancetype":{"name":"u1.micro","revisionName":""}}}'
+
+./cluster-up/kubectl.sh wait vmis/alpine --for=condition=HotMemoryChange
+
+while ! ./cluster-up/kubectl.sh get virtualmachineinstancemigrations -l kubevirt.io/vmi-name=alpine; do
+  sleep 1
+done 
+
+./cluster-up/kubectl.sh wait virtualmachineinstancemigrations -l kubevirt.io/vmi-name=alpine --for=jsonpath='{.status.phase}'="Succeeded"
+
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.status.memory.guestAtBoot}'="512Mi"
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.status.memory.guestRequested}'="1Gi"
+./cluster-up/kubectl.sh wait vmis/alpine --for=jsonpath='{.status.memory.guestCurrent}'="1Gi"
+```
+
+## Scalability
+
+N/A
+
+## Update/Rollback Compatibility
```

Comment:
Lets say I've live-updated an instance-type of a running VM from `u1.nano` to `u1.micro`, what will happen if the live-update fails for some reason and I already deleted the `u1.nano` instance type CR?